### PR TITLE
Add handlers module & allow users to set up their own handlers

### PR DESCRIPTION
**Pull request by Marahin**
* Added handlers module,
* handlers interface that allows users to define custom handlers, 
* resource handlers matching so users can enforce using a specific handler for a given `relationship_key`.

## What is this about

There are situations where you want to have more control over how is a given relationship key handled. In some of the previous issues there was a discussion on how far can we go with assumptions about the data type / relation type (1-n, 1-1, ...) based on linguistics.

Our primitive solution (based on pluralization/singularization of the `relationship_key`) is doing fine in most cases, but at some point you will want to have some kind of customizations. It was the main reason to allow users to have their own, custom handlers. 

In order to understand what is the concept behind the handlers usage, see this file: https://github.com/visualitypl/jsonapi_parameters/pull/28/files#diff-66523cf18661e0669fba54b1fa735a97

## TODO

* Better documentation - considering to create Wiki pages for this project. Handlers are complex to understand.
* Integration test covering actually using a custom handler in a Rails app (`./spec/app`)

### CODE REVIEW

File: lib/jsonapi_parameters/handlers.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          raise NotImplementedError.new(
+            'handler_key does not match any registered handlers'
+          )
+        end
+
+        resource_handlers[resource_key.to_sym] = handler_key.to_sym
+      end
+
+      def reset_handlers!
+        @handlers = {}.merge(DEFAULT_HANDLER_SET)
```
Comment (pstrzalk) [inline lib/jsonapi_parameters/handlers.rb]: `DEFAULT_HANDLER_SET.dup` would be more explicit

File: spec/lib/jsonapi_parameters/handlers_spec.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+      horse_handler = ->(relationship_key, _, _) { ["#{relationship_key}_id".to_sym, nil] }
+
+      JsonApi::Parameters::Handlers.add_handler(:handle_plural_nil_as_belongs_to_nil, horse_handler)
+      JsonApi::Parameters::Handlers.set_resource_handler(:scissors, :handle_plural_nil_as_belongs_to_nil)
+
+      result = trs.jsonapify(params)
+
+      expect(result).to eq(
+        user: { first_name: 'John', id: '666', scissors_id: nil }
```
Comment (pstrzalk) [inline spec/lib/jsonapi_parameters/handlers_spec.rb]: 👍  Exactly what we needed

File: lib/jsonapi_parameters/default_handlers/to_many_relation_handler.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
+module JsonApi
+  module Parameters
+    module Handlers
+      module DefaultHandlers
+        class ToManyRelationHandler < BaseHandler
+          include ActiveSupport::Inflector
+
+          def handle
```
[the remark this answers is not in the document]
Comment (Marahin) [reply lib/jsonapi_parameters/default_handlers/to_many_relation_handler.rb]: I agree!

File: spec/lib/jsonapi_parameters/handlers_spec.rb [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+end
+
+describe JsonApi::Parameters::Handlers do
+  before(:each) do
+    described_class.reset_handlers!
+  end
+
+  describe 'reset_handlers!' do
+    it 'sets all the handlers to the default handlers' do
+      handler = described_class.add_handler(:test, -> { puts 'Hello!' })
```
[the remark this answers is not in the document]
Comment (Marahin) [reply spec/lib/jsonapi_parameters/handlers_spec.rb]: The point of this test is to ensure the result of `reset_handlers!` which is a default handler set. 

There is another test for `add_handler` method, but additionally I will make it a bit more rich and verbose about what is it actually testing.

File: spec/lib/jsonapi_parameters/handlers_spec.rb [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+          attributes: {
+            first_name: 'John'
+          },
+          relationships: {
+            scissors: { data: nil }
+          }
+        }
+      }
+
+      horse_handler = ->(relationship_key, _, _) { ["#{relationship_key}_id".to_sym, nil] }
```
[the remark this answers is not in the document]
Comment (Marahin) [reply spec/lib/jsonapi_parameters/handlers_spec.rb]: Forgot about it, good catch!

File: lib/jsonapi_parameters/default_handlers/to_many_relation_handler.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                included_object[:attributes].merge(id: related_id)
+              else
+                relationship.dig(:id)
+              end
+            end
+
+            # We may have smells in our value array as `with_inclusion` may have been changed at some point
+            # but not in the beginning.
+            # Because of that we should clear it and make sure the results are unified (e.g. array of ids)
+            unless with_inclusion
```
[the remark this answers is not in the document]
Comment (Marahin) [reply lib/jsonapi_parameters/default_handlers/to_many_relation_handler.rb]: I guess I just had an idea about keeping consistency within the returned data but now I am no longer sure what is this about.

Good catch, removing it. Maybe one day I will get enlightened why I wanted to actually map through the values.

File: spec/lib/jsonapi_parameters/handlers_spec.rb [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+      described_class.reset_handlers!
+
+      expect(described_class.resource_handlers).to eq({})
+    end
+  end
+
+  describe 'custom handlers' do
+    it 'allows to add a custom handler' do
+      expect(described_class.add_handler(:test, -> { puts 'Hello!' })).to be_an_instance_of(Proc)
```
[the remark this answers is not in the document]
Comment (Marahin) [reply spec/lib/jsonapi_parameters/handlers_spec.rb]: Correct. I have enriched this test to be more verbose and to check whether it has been actually added to `@handlers` variable.

File: lib/jsonapi_parameters/handlers.rb (hunk 1 above)
Comment (Marahin) [reply lib/jsonapi_parameters/handlers.rb]: I knew there was an elegant way to do this! Thanks! 🥇

**Pull request merged**